### ISSUE TITLE
Using file watching to update /var/cache/tang is unreliable

### ISSUE THREAD

**Issue by mvollmer** [latchset/tang#23]
Run this script:

```
gensig() {
  jwe=`jose jwk gen -i '{"alg":"ES512"}'`
  sig=`echo "$jwe" | jose jwk thp -i-`
  echo $sig
  echo "$jwe" > /var/db/tang/$sig.jwk
}

rm -rf /var/db/tang/* /var/cache/tang/*
gensig
gensig
gensig
gensig
gensig
systemctl start tangd-update

sleep 1
echo ====
cat /var/cache/tang/default.jws | jose fmt -j- -g payload -y -o- | jose jwk thp -i-
```

It puts five signing keys into /var/db/tang, but default.jws never contains more than three of them for me.  Here is a typical output:

```
ftVX1t2ls9G2X3SNfFXrGtFWq7Y
B7U_kOUlaes4yuPtiAwUqzCL2Lc
OChmt6EW-jDmEnkHPoRcfQGHfDA
wVT1AuHo2ZIcbbSzbZ25YavgPbQ
dtkh3lW40H2bkJkOlEBw6N952Po
====
B7U_kOUlaes4yuPtiAwUqzCL2Lc
ftVX1t2ls9G2X3SNfFXrGtFWq7Y
```

Looking at the journal, tangd-udpate is only started once for each run of the script, probably triggered by the `rm`.  The explicit start doesn't do anything, probably because it is still running.

Changing `systemctl start tangd-update` to `restart` makes it all work.

This also happens for me with just tangd-keygen.  In our test VMs, the "deriveKey" key that is generated last by tangd-keygen is almost always missing from the cache when we start tangd.socket.

So, for reliability, we should tell people to run `systemctl restart tangd-update` after making changes to `/var/db/tang`.  Which kinda makes `tangd-update.path` superfluous, no?

**Comment by mvollmer** [comment]
> we should tell people to run systemctl restart tangd-update

And tangd-keygen should do that, too.

**Comment by npmccallum** [comment]
If this can't be made to work, we should probably move the signature generation into Tang itself.

**Comment by oleshii** [comment]
Hello Nathaniel, 
How are you ? Happy Merry Christmas !
Currently I am working over this task. I got the problem.
I built and installed jose and tangd manually. After the typing
sudo systemctl enable tangd.socket --now
I saw this message in the system journal: 
'Failed to start Tang Server key update script. 'Invalid input!'
My research showed that tangd-keygen script had only one argument on start:
/usr/lib/systemd/system/tangd-keygen.service
ExecStart=/usr/local/libexec/tangd-keygen /usr/local/var/db/tang
Requires=tangd-update.path
But in the real life second argument is necessary
What is more argument should I assign to work this script properly ?

**Comment by oleshii** [comment]
No problem. I was able made sence myself.

**Comment by oleshii** [comment]
Hello Nathaniel,
Hello Daniel,
I can not reproduce it more.
I decided some access rights are affect.
I built projects manually and install them too.
I.E. when they are installed manually they used
other working directories, /usr/local/var/db/tang/ and /usr/local/var/cache/tang/
namely. I assigned other access rights for my own convinience only.
Now they are: ls -al /usr/local/var/
drwx------   4 oleshii oleshii 4096 дек 25 15:51 .
drwxr-xr-x. 14 oleshii oleshii 4096 дек 25 15:51 ..
drwxr-xr-x   3 oleshii oleshii 4096 дек 25 15:51 cache
drwx------   3 oleshii oleshii 4096 дек 25 15:51 db
Now the problem did not reproduced.
I can suggest to the user to change the access rights fro /var/cache and /var/db, or execute the script over than superuser environment.

**Comment by cedricbu** [comment]
Hi,
I believe that RHEL-8 fixed this issue [as well as #24] via moving the key generation to tang & remove the cache directory entirely : see https://bugzilla.redhat.com/show_bug.cgi?id=1679186
However, the corresponding patch (https://bugzilla.redhat.com/attachment.cgi?id=1625528) does not seem to have been pushed upstream.

I have not tried the patch, but if it resolves the issue, it would help other distros to benefit from it ☺
